File: test/corpus/control/for_reduce.exs
defmodule RustQ.Corpus.Control.ForReduce do
  @moduledoc "Fallible for/reduce lowering."

  use RustQ.Meta

  alias RustQ.Type, as: R

  @spec validate(R.vec(R.u32())) :: R.nif_result(R.unit())
  defrust validate(values) do
    for value <- values, reduce: :ok do
      :ok ->
        if value == 0 do
          {:error, badarg()}
        else
          :ok
        end
    end
  end
end

File: test/corpus/control/for_reduce.rs
fn validate(values: Vec<u32>) -> NifResult<()> {
    {
        let mut __rustq_reduce = Ok(());
        for value in values {
            __rustq_reduce = match __rustq_reduce {
                Ok(()) => {
                    if value == 0 {
                        Err(rustler::Error::BadArg)
                    } else {
                        Ok(())
                    }
                }
                __rustq_reduce_value => __rustq_reduce_value,
            };
        }
        __rustq_reduce
    }
}
